### DISCUSSION TITLE
PIN assignments and digital PWM

### DISCUSSION THREAD

**Discussion by coderatchance** [SpenceKonde/megaTinyCore#1057] [Library discussion]
Hello all,

I am getting out of analog PWM ports, hence I am looking at how to enable and use the digital PWM ports.
For example, when looking at the Attinyx_06.gif for the At tiny 406  the only extra PWM is available at the PB2 , namely digital PWM, however, I am not really sure how to use it with the Arduino code? Any ideas, examples?

Best.

**Comment by hmeijdam** [comment]
Here an example

```
void setup() {
}

void loop() {
  static uint8_t dutycycle;
  analogWrite(PIN_PB2, dutycycle);
  dutycycle++;
  delay(5);
}
```

O, and I get the feeling that you think that you must use analogWrite with analog pins. Analog pins are only required for analogRead, since the analog pins can be used by the ADC. They should have called "analogWrite" "pwmWrite". Then nobody has to be confused.

**Reply by coderatchance** [reply]
I am coming back to this issue.
Tried the given example, and cannot get the led dim nicely, it only works when 0 or 255:
`digitalWrite(PIN_PB5, 100);` the led is off?

Of course on this pins (PB5, PB4, and PB3) the TCA0 can be used but when 'swapped' , sorry I am under influence of swapping the serial ports, anyway, I would like to turn on all of the possible PWMs on the 3216 and the first three (PA3m PA4 and PA5) work ok.

Any ideas ?

**Reply by hmeijdam** [reply]
?? I see you are using `digitalWrite(PIN_PB5, 100)` 

digitalWrite will never give you any PWM. You need analogWrite for that.

**Reply by coderatchance** [reply]
Indeed, no matter what is used I do not get the PWM, I guess I need to 'swap' the port, timer WO2, WO1, WO3 ?
How can this be done? 

" 4. Alternative pin location. For selecting an alternative pin location, refer to the PORTMUX section."

**Reply by hmeijdam** [reply]
If you want to program registers directly according the datasheet, you better not use the Arduino functions. It's one or the other, unless you know exactly what the Arduino functions do with the registers.
See https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/TakingOverTCA0.md

**Comment by coderatchance** [comment]
Thanks for the link, explanation.
Looking at the 3216 pinouts According to the Arduino conventional pin numbering.

I am after using the 8 PWMs , and if I am not wrong, those can be manipulated with the analogWrite() in the Arduino IDE on all of the 8 pins with the tilde? That is: 0, 1, 7, 8, 9, 10, 11 and 16?In this case, if I want to use the TWI, namely the SDA and SCL,  then the alternative pins have to be used, like 14 and 15? 

Not sure how to tackle this.However, at some times along those in assignments, the RXD and TXD are as well needed. 
I guess changing during the program routine between the TXD and PWM on the pin 7 is not that a big problem? Is this the right way to look at it? 
In addition, such a mapping could as well be used on the 1616?

**Reply by hmeijdam** [reply]
See the pinswap_via_swap.ino example sketch that's part of your installation.

and read the [wire](https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/libraries/Wire/README.md) docs

**Reply by coderatchance** [reply]
Cheers! 
Using the following and works like a charm...

```
  Wire.swap(1);
  Wire.begin(); //Join i2c bus
  Wire.setClock(100000); //Run the I2C bus at speed ?  In your case max 400kHz, 100kHz , 10kHz min!
```

**Reply by coderatchance** [reply]
Ok, how fast do you think the serial can be put back once it is removed for the PWM?
For example, the serial TXD is used for printing out, however, at points this pins is as well used for the PWM, at that time the serial is not need, hence it will be removed and brought back...
So the question is, when this occurs, the serial will have to end and when the PWM is done, it will have to come back, and so on...
Does this make any sense?

**Reply by hmeijdam** [reply]
That's impossible to say, as it depends on your code. What you can do is do 1.000.000 times serial.begin() and serial.end() in a loop and not much else. and measure the time that takes.

**Reply by coderatchance** [reply]
You are right, even the idea is too far fetched, will try to setup the TCD0, namely WOC, WOD, if not successful, will try the software PWM...

**Reply by coderatchance** [reply]
Did some reading on PWM, namely [analogWrite() and configuring TCD](https://github.com/SpenceKonde/megaTinyCore/blob/master/megaavr/extras/Ref_TCD.md) , but the more I read the more I get confused, so, please help me first understand the following before I delve in:

ATTiny3216-The pins (Arduino names with tilde) can be used with the `analogWrite(pin_no, 0-255)` namely the PWM with the TCA0 and TCD0 ;

1. pin 0,
2. pin 1 
3. pin 7 
4. pin 8, 
5. pin 9, 
6. pin 10 (TCD0  ?)
8. pin 11 (TCD0 ?)
9. pin 16

Tested them with the led and analogWrite(); and it works well. 
However, I will need the pin 7 for the TXD, because I am using the alternative serial pins, for the SCL, SDA.
So , the options for another PWM pin are to use the software PWM ?
Have the simple method in mind, however, do we know for a library that could be used?

Best.
